### PR TITLE
Add DPO and ORPO preference data preprocessing pipeline utils

### PR DESCRIPTION
**Pull request by igorts-git**
# Description

To simplify code review  I am splitting the Tunix-based DPO implementation into smaller PRs.
This one adds the data reading processing required by DPO.

The classic DPO inputs consist of three data columns: ["prompt", "chosen_response", "rejected_response"]. 
However, some DPO datasets use a two-column format where the prompt is the prefix to the choosen and rejected strings.
When a 2-column dataset is used our implementation extracts the common prefix into the "prompt" field that is then fed into the model separately.
The column names in the dataset can wary, for example ["input", chosen", "rejected"]. Our implementation allows the user to supply the dataset column names via the `train_data_columns` and `eval_data_columns` parameters.

Tunix requires left-padded prompt and right-padded responses. Our code implements this padding (and truncation if needed) it also provides Tunix with the corresponding masks.

NOTE: once this PR is merged the legacy DPO will stop working correctly. The follow up PRs will enable Tunix-based DPO.

Caveat: This PR only adds support for HuggingFace datasets, while the legacy DPO implementation supported HuggingFace, TFDS and Grain. This is on-par with our SFT implementation. We need to discuss the priority of supporting TFDS and Grain in post-training.

# Tests

Added unit tests. Ran DPO/ORPO and performed logits comparison against the legacy implementation.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/input_pipeline/dpo_utils.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      raise KeyError(
+          f"Column '{e.args[0]}' not found in the dataset. "
+          f"Expected columns: {self.data_column_names}. "
+          f"Available columns: {list(element.keys())}. "
+          "Please verify that 'train_data_columns' and 'eval_data_columns' match your dataset."
+      ) from e
+
+    # 2. Padding and Masking
+    max_prompt_length = self.max_target_length // 2
+    max_response_length = self.max_target_length - max_prompt_length
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/input_pipeline/dpo_utils.py:68]: added "max_prompt_length" parameter

File: src/maxtext/input_pipeline/dpo_utils.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    Padding can be done on the left or right. Trimming is only on the right.
+    """
+    x = np.asarray(x)
+    pad_amount = max(length - x.shape[0], 0)
+    if left:
+      pad_width = ((pad_amount, 0),)
+    else:
+      pad_width = ((0, pad_amount),)
+    return np.pad(x[:length], pad_width, constant_values=self.pad_id).astype(np.int32)
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/input_pipeline/dpo_utils.py]: applied the recommended patch (with a small modification)

File: src/maxtext/input_pipeline/dpo_utils.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        full_chosen = element[self.data_column_names[0]]
+        full_rejected = element[self.data_column_names[1]]
+
+        # Find common prefix length
+        i = 0
+        for i, (c, r) in enumerate(zip(full_chosen, full_rejected)):
+          if c != r:
+            break
+        input_ids = full_chosen[:i]
+        chosen_ids = full_chosen[i:]
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/input_pipeline/dpo_utils.py]: applied the recommended patch

File: tests/post_training/unit/dpo_data_processing_test.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+  def test_two_column_prefix_extraction(self):
+    """Verify common prefix extraction for 2-column datasets."""
+    # The column names will be remappend into "chosen" and "rejected"
+    prep = dpo_utils.DPOTunixPrep(pad_id=self.pad_id, max_target_length=21, data_column_names=("liked", "disliked"))
+    sample = {
+        "liked": np.array([1, 2, 3, 10, 11]),
+        "disliked": np.array([1, 2, 3, 20, 21, 22]),
+    }
+    output = prep.map(sample)
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply tests/post_training/unit/dpo_data_processing_test.py:80]: added a test case

File: src/maxtext/input_pipeline/dpo_utils.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import dataclasses
+import grain.python as grain
+import numpy as np
+
+
+@dataclasses.dataclass
+class DPOTunixPrep(grain.MapTransform):
+  """Prepares DPO data for Tunix.
+  Renames input columns, extracts common prefix if needed, generates masks,and performs
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/input_pipeline/dpo_utils.py]: reformatted

File: src/maxtext/configs/types.py [hunk 6]
```diff
@@ -1201,12 +1217,23 @@ class OlmoGrainDataset(BaseModel):
...
 
+class DPO(BaseModel):
+  """Configuration for DPO and ORPO preference optimization algorithms."""
+
+  algo: Literal["dpo", "orpo"] = Field("dpo", description="Alignment algorithm to use.")
+  dpo_beta: float = Field(0.1, description="Beta parameter for DPO.")
+  orpo_lambda: float = Field(0.1, description="Weight for preference loss in ORPO.")
+  dpo_label_smoothing: float = Field(0.0, ge=0.0, le=1.0, description="Label smoothing for DPO.")
+  max_prompt_length: int | None = Field(
+      None,
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/configs/types.py:1217]: applied with "gt=0" instead of the suggested "ge=0"

File: src/maxtext/input_pipeline/dpo_utils.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    """Pads or trims an array to a specific length.
+
+    When left=True (for prompts), trims from the left to keep the suffix (closest context).
+    When left=False (for responses), trims from the right to keep the prefix.
+    """
+    x = np.asarray(x)
+    pad_amount = max(length - x.shape[0], 0)
+    if left:
+      pad_width = ((pad_amount, 0),)
+      x_trimmed = x[-length:]
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/input_pipeline/dpo_utils.py:104]: Instead added an assertion in a different place to make sure that length is never 0

File: src/maxtext/configs/base.yml [hunk 8]
```diff
@@ -680,8 +680,12 @@ global_rampup_samples: 500
 
 # direct preference optimization (DPO)
 use_dpo: False
-dpo_label_smoothing: 0.0
-dpo_beta: 0.1
+dpo:
```
Comment (SurbhiJainUSC) [inline src/maxtext/configs/base.yml]: Do we need to add these DPO configs to base.yml too?

File: src/maxtext/input_pipeline/dpo_utils.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+"""DPO specific input pipeline utilities."""
+
+import dataclasses
+import grain.python as grain
+import numpy as np
+
+
+@dataclasses.dataclass
+class DPOTunixPrep(grain.MapTransform):
```
Comment (SurbhiJainUSC) [inline src/maxtext/input_pipeline/dpo_utils.py]: Please rename it to `DPODataFormatting` or something similar

File: tests/post_training/unit/dpo_data_processing_test.py [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+    )
+    # 'rejected' column is missing
+    sample = {"input": np.array([1, 2]), "chosen": np.array([3, 4])}
+    with self.assertRaisesRegex(KeyError, "Column 'rejected' not found in the dataset"):
+      prep.map(sample)
+
+
+@pytest.mark.external_training
+class TestDPOPipelineIntegration(unittest.TestCase):
+  """End-to-end DPO pipeline integration tests."""
```
Comment (SurbhiJainUSC) [inline tests/post_training/unit/dpo_data_processing_test.py]: Is this really an integ test?

File: src/maxtext/configs/post_train/dpo.yml [hunk 11]
```diff
@@ -24,8 +30,6 @@ hf_eval_split: 'test'
 
 gradient_clipping_threshold: 10.0
 learning_rate: 5.0e-7
-dpo_label_smoothing: 0.0
-dpo_beta: 0.1
```
Comment (gagika) [inline src/maxtext/configs/post_train/dpo.yml:28]: will this break older DPO code, if so should we just leave as it is for now, perhpas comment that used for older DPO?

**Review by gagika — APPROVED** [review]

File: src/maxtext/configs/post_train/dpo.yml (hunk 11 above)
Comment (igorts-git) [reply src/maxtext/configs/post_train/dpo.yml:28]: Unfortunately, this PR breaks the older DPO code, not just due to the configs, but also in how the dataset is loaded. Once this and the next PR in the series is merged, I plan to follow up with a PR that completely deletes the legacy DPO implementation.

File: src/maxtext/configs/base.yml (hunk 8 above)
Comment (igorts-git) [reply src/maxtext/configs/base.yml]: Sorry, I don't fully understand the comment. Are you suggesting to remove these configs from base.yml?

File: src/maxtext/input_pipeline/dpo_utils.py (hunk 9 above)
Comment (igorts-git) [reply src/maxtext/input_pipeline/dpo_utils.py]: Renamed the class.

File: tests/post_training/unit/dpo_data_processing_test.py (hunk 10 above)
Comment (igorts-git) [reply tests/post_training/unit/dpo_data_processing_test.py]: you're right, it does not fall into the "integration test" category. rephrased.

File: src/maxtext/configs/types.py [hunk 12]
```diff
@@ -2887,6 +2903,16 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
           raise ValueError("For multimodal SFT, `sft_train_on_completion_only` must be True.")
         if self.packing:
           raise ValueError("For multimodal SFT, `packing` is not yet supported.")
+    if self.use_dpo:
+      if self.packing:
+        raise ValueError("For DPO/ORPO, `packing` is not supported.")
+      if self.dpo.max_prompt_length is not None and self.dpo.max_prompt_length > self.max_target_length:
```
Comment (aireenmei) [inline src/maxtext/configs/types.py]: In the case of max_prompt_length == max_target_length, it will cause max_response_length=0 error in DPODataFormatting, should we guard it here?

File: src/maxtext/configs/base.yml (hunk 8 above)
Comment (A9isha) [reply src/maxtext/configs/base.yml]: most likely these `dpo` related parameters ended up being in `base.yml` due to historical reasons. if we have them in the `dpo.yaml` and we have separate yml files in the `configs/post_train` directory then it makes sense to remove them from `base.yml` since the ones in `base.yml` are meant to be shareable across multiple use-cases.

**Review by aireenmei — APPROVED** [review]
LGTM. But please note that both tfds and hf pipelines are planned for deprecation. There's already support for [SFT](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/input_pipeline/grain_data_processing.py#L316) and [DPO](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/input_pipeline/grain_data_processing.py#L258) in Grain pipeline. It would be great if we can have follow up changes to enable the same support in Grain pipeline

**Review by A9isha — APPROVED** [review]
thanks Igor

Approved barring the one comment

File: src/maxtext/configs/types.py (hunk 12 above)
Comment (igorts-git) [reply src/maxtext/configs/types.py]: Updated the guard here. FYI, There is a slightly more comprehensive assertion in dpo_utils.py that can trigger in a few more edge cases.

File: src/maxtext/configs/base.yml (hunk 8 above)
Comment (igorts-git) [reply src/maxtext/configs/base.yml]: OK, thanks for the details. I removed the "dpo:" section from `base.yml`
